src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs

```diff
@@ -2476,6 +2487,772 @@ static void Main()
+                Diagnostic(ErrorCode.ERR_CannotInferDelegateType, "(ArgIterator x) => { }").WithLocation(8, 18));
+        }
+
+        [WorkItem(55217, "https://github.com/dotnet/roslyn/issues/55217")]
+        [Fact]
+        public void SynthesizedDelegateTypes_13()
+        {
+            var source =
+@"ref struct S<T> { }
+class Program
+{
+    static void F1(int x, S<int> y) { }
+    static S<T> F2<T>() => throw null;
+    static void Main()
+    {
+        var d1 = F1;
+        var d2 = F2<object>;
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (8,18): error CS8917: The delegate type could not be inferred.
+                //         var d1 = F1;
+                Diagnostic(ErrorCode.ERR_CannotInferDelegateType, "F1").WithLocation(8, 18),
+                // (9,18): error CS8917: The delegate type could not be inferred.
+                //         var d2 = F2<object>;
+                Diagnostic(ErrorCode.ERR_CannotInferDelegateType, "F2<object>").WithLocation(9, 18));
+        }
+
+        [Fact]
+        public void SynthesizedDelegateTypes_14()
+        {
+            var source =
+@"class Program
+{
+    static ref void F() { }
+    static void Main()
+    {
+        var d1 = F;
+        var d2 = (ref void () => { });
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (3,16): error CS1547: Keyword 'void' cannot be used in this context
+                //     static ref void F() { }
+                Diagnostic(ErrorCode.ERR_NoVoidHere, "void").WithLocation(3, 16),
+                // (6,18): error CS8917: The delegate type could not be inferred.
+                //         var d1 = F;
+                Diagnostic(ErrorCode.ERR_CannotInferDelegateType, "F").WithLocation(6, 18),
+                // (7,19): error CS8917: The delegate type could not be inferred.
+                //         var d2 = (ref void () => { });
+                Diagnostic(ErrorCode.ERR_CannotInferDelegateType, "ref void () => { }").WithLocation(7, 19),
+                // (7,23): error CS1547: Keyword 'void' cannot be used in this context
+                //         var d2 = (ref void () => { });
+                Diagnostic(ErrorCode.ERR_NoVoidHere, "void").WithLocation(7, 23));
+        }
+
+        private static void VerifyLocalDelegateType(SemanticModel model, VariableDeclaratorSyntax variable, string expectedLocal, string expectedInvokeMethod)
+        {
+            var local = (ILocalSymbol)model.GetDeclaredSymbol(variable)!;
+            Assert.Equal(expectedLocal, local.ToTestDisplayString());
+            var delegateType = ((INamedTypeSymbol)local.Type);
+            Assert.Equal(Accessibility.Internal, delegateType.DeclaredAccessibility);
+            Assert.Equal(expectedInvokeMethod, delegateType.DelegateInvokeMethod.ToTestDisplayString());
+        }
```

Comment:
Let's include a test with more types: dynamic, tuples with names, nint or reference types with nullability.

Reply:
To add to the list, arrays of function pointer types where the parameters/return type of the function pointer have refkinds.

Reply:
I also think we should have public API verification, ie what are we returning from `GetSymbolInfo` or `GetTypeInfo` for these synthesized symbols.